File: library/src/commonMain/kotlin/com/doordeck/multiplatform/sdk/api/responses/PlatformResponses.kt
package com.doordeck.multiplatform.sdk.api.responses

import kotlinx.serialization.SerialName
import kotlinx.serialization.Serializable
import kotlinx.serialization.json.JsonClassDiscriminator
import kotlin.js.JsExport

@JsExport
@Serializable
class ApplicationResponse(
    val applicationId: String,
    val name: String,
    val lastUpdated: Double? = null,
    val owners: Array<String>? = null,
    val corsDomains: Array<String>? = null,
    val authDomains: Array<String>? = null,
    val logoUrl: String? = null,
    val privacyPolicy: String? = null,
    val mailingAddress: String? = null,
    val companyName: String? = null,
    val supportContact: String? = null,
    val appLink: String? = null,
    val slug: String? = null,
    val emailPreferences: EmailPreferencesResponse,
    val authKeys: Container<AuthKeyResponse>,
    val oauth: OauthResponse? = null,
    val isDoordeckApplication: Boolean? = null
)

@JsExport
@Serializable
@JsonClassDiscriminator("kty")
sealed interface AuthKeyResponse {
    val kid: String
    val use: String
    val alg: String?
    val ops: Array<String>?
    val x5u: String?
    val x5t: String?
    val x5t256: String?
    val x5c: Array<String>?
    val exp: Int?
    val nbf: Int?
    val iat: Int?
}

@JsExport
@Serializable
@SerialName("RSA")
class RsaKeyResponse(
    override val use: String,
    override val kid: String,
    override val alg: String? = null,
    @SerialName("key_ops")
    override val ops: Array<String>? = null,
    override val x5u: String? = null,
    override val x5t: String? = null,
    @SerialName("x5t#S256")
    override val x5t256: String? = null,
    override val x5c: Array<String>? = null,
    override val exp: Int? = null,
    override val nbf: Int? = null,
    override val iat: Int? = null,
    val e: String,
    val n: String
): AuthKeyResponse

@JsExport
@Serializable
@SerialName("EC")
class EcKeyResponse(
    override val use: String,
    override val kid: String,
    override val alg: String? = null,
    @SerialName("key_ops")
    override val ops: Array<String>? = null,
    override val x5u: String? = null,
    override val x5t: String? = null,
    @SerialName("x5t#S256")
    override val x5t256: String? = null,
    override val x5c: Array<String>? = null,
    override val exp: Int? = null,
    override val nbf: Int? = null,
    override val iat: Int? = null,
    val crv: String,
    val x: String,
    val y: String
): AuthKeyResponse

@JsExport
@Serializable
@SerialName("OKP")
class Ed25519KeyResponse(
    override val use: String,
    override val kid: String,
    override val alg: String? = null,
    @SerialName("key_ops")
    override val ops: Array<String>? = null,
    override val x5u: String? = null,
    override val x5t: String? = null,
    @SerialName("x5t#S256")
    override val x5t256: String? = null,
    override val x5c: Array<String>? = null,
    override val exp: Int? = null,
    override val nbf: Int? = null,
    override val iat: Int? = null,
    val d: String? = null,
    val crv: String,
    val x: String
): AuthKeyResponse

@JsExport
@Serializable
class EmailPreferencesResponse(
    val senderEmail: String? = null,
    val senderName: String? = null,
    val primaryColour: String,
    val secondaryColour: String,
    val onlySendEssentialEmails: Boolean? = null,
    val callToAction: EmailCallToActionResponse? = null,
)

@JsExport
@Serializable
class EmailCallToActionResponse(
    val actionTarget: String,
    val headline: String,
    val actionText: String
)

@JsExport
@Serializable
class OauthResponse(
    val authorizationEndpoint: String,
    val clientId: String,
    val grantType: String
)

@JsExport
@Serializable
class ApplicationOwnerDetailsResponse(
    val userId: String,
    val email: String,
    val displayName: String? = null,
    val orphan: Boolean,
    val foreign: Boolean
)

@JsExport
@Serializable
class GetLogoUploadUrlResponse(
    val uploadUrl: String
)

File: doordeck-sdk/src/mingwMain/resources/csharp/Model/Responses/PlatformResponses.cs
using System.Text.Json.Serialization;

namespace Doordeck.Headless.Sdk.Model.Responses;

public class ApplicationResponse
{
    public required Guid ApplicationId { get; set; }
    public required string Name { get; set; }
    public DateTime? LastUpdated { get; set; }
    public List<Guid>? Owners { get; set; }
    public List<Uri>? CorsDomains { get; set; }
    public List<Uri>? AuthDomains { get; set; }
    public Uri? LogoUrl { get; set; }
    public Uri? PrivacyPolicy { get; set; }
    public string? MailingAddress { get; set; }
    public string? CompanyName { get; set; }
    public Uri? SupportContact { get; set; }
    public Uri? AppLink { get; set; }
    public string? Slug { get; set; }
    public required EmailPreferencesResponse EmailPreferences { get; set; }
    public required Dictionary<string, AuthKeyResponse> AuthKeys { get; set; }
    public OauthResponse? Oauth { get; set; }
    public bool? IsDoordeckApplication { get; set; }
}

public interface IAuthKeyResponse;

[JsonPolymorphic(TypeDiscriminatorPropertyName = "kty")]
[JsonDerivedType(typeof(RsaKeyResponse), "RSA")]
[JsonDerivedType(typeof(EcKeyResponse), "EC")]
[JsonDerivedType(typeof(Ed25519KeyResponse), "OKP")]
public abstract class AuthKeyResponse : IAuthKeyResponse;

public class RsaKeyResponse : AuthKeyResponse
{
    public required string Kty { get; set; }
    public required string Kid { get; set; }
    public required string Use { get; set; }
    public string? Alg { get; set; }
    [JsonPropertyName("key_ops")]
    public List<string>? Ops { get; set; }
    public string? X5U { get; set; }
    public string? X5T { get; set; }
    [JsonPropertyName("x5t#S256")]
    public string? X5T256 { get; set; }
    public List<string>? X5C { get; set; }
    public int? Exp { get; set; }
    public int? Nbf { get; set; }
    public int? Iat { get; set; }
    public required string E { get; set; }
    public required string N { get; set; }
}

public class EcKeyResponse : AuthKeyResponse
{
    public required string Kty { get; set; }
    public required string Kid { get; set; }
    public required string Use { get; set; }
    public string? Alg { get; set; }
    [JsonPropertyName("key_ops")]
    public List<string>? Ops { get; set; }
    public string? X5U { get; set; }
    public string? X5T { get; set; }
    [JsonPropertyName("x5t#S256")]
    public string? X5T256 { get; set; }
    public List<string>? X5C { get; set; }
    public int? Exp { get; set; }
    public int? Nbf { get; set; }
    public int? Iat { get; set; }
    public required string Crv { get; set; }
    public required string X { get; set; }
    public required string Y { get; set; }
}

public class Ed25519KeyResponse : AuthKeyResponse
{
    public required string Kty { get; set; }
    public required string Kid { get; set; }
    public required string Use { get; set; }
    public string? Alg { get; set; }
    [JsonPropertyName("key_ops")]
    public List<string>? Ops { get; set; }
    public string? X5U { get; set; }
    public string? X5T { get; set; }
    [JsonPropertyName("x5t#S256")]
    public string? X5T256 { get; set; }
    public List<string>? X5C { get; set; }
    public int? Exp { get; set; }
    public int? Nbf { get; set; }
    public int? Iat { get; set; }
    public required string Crv { get; set; }
    public required string X { get; set; }
}

public class EmailPreferencesResponse
{
    public string? SenderEmail { get; set; }
    public string? SenderName { get; set; }
    public required string PrimaryColour { get; set; }
    public required string SecondaryColour { get; set; }
    public bool? OnlySendEssentialEmails { get; set; }
    public required EmailCallToActionResponse? CallToAction { get; set; }
}

public class EmailCallToActionResponse
{
    public required Uri ActionTarget { get; set; }
    public required string Headline { get; set; }
    public required string ActionText { get; set; }
}

public class OauthResponse
{
    public required string AuthorizationEndpoint { get; set; }
    public required string ClientId { get; set; }
    public required GrantType GrantType { get; set; }
}

public class ApplicationOwnerDetailsResponse
{
    public Guid UserId { get; set; }
    public required string Email { get; set; }
    public string? DisplayName { get; set; }
    public required bool Orphan { get; set; }
    public required bool Foreign { get; set; }
}

public class GetLogoUploadUrlResponse
{
    public required Uri UploadUrl { get; set; }
}